src/test/java/Assignment2/StudentServiceTest.java

```diff
@@ -0,0 +1,228 @@
+package Assignment2;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class StudentServiceTest {
+
+  @Test
+  void checkIfNameEmpty() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("", "12", "abs scs", "12", "A B C D"));
+  }
+
+  @Test
+  void checkIfRollNoEmpty() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("abc", "", "abs scs", "12", "A B C D"));
+  }
+
+  @Test
+  void checkIfAddressEmpty() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("abc", "12", "", "12", "A B C D"));
+  }
+
+  @Test
+  void checkIfAgeEmpty() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("abc", "12", "abs scs", "", "A B C D"));
+  }
+
+  @Test
+  void checkIfCourseEmpty() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("abc", "12", "abs scs", "12", ""));
+  }
+
+  @Test
+  @DisplayName("When name format is wrong")
+  void checkInputName() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("abc2", "12", "x y z", "21", "A B C D"));
+  }
+
+  @Test
+  @DisplayName("When rollNo format is wrong")
+  void checkInputRollNo() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("abc", "2F", "x y z", "21", "A B C D"));
+  }
+
+  @Test
+  @DisplayName("When Age format is wrong")
+  void checkInputAge() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("abc", "12", "x y z", "2A", "A B C D"));
+  }
+
+  @Test
+  @DisplayName("When Course format is wrong")
+  void checkInputCourse() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("abc", "12", "x y z", "21", "A B C 1 2 34"));
+  }
+
+  @Test
+  void checkIfRepeatCourse() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("abc", "12", "abs scs", "12", "A A C D"));
+  }
+
+  @Test
+  void checkIfAtleast4Course() {
+    StudentService studentService = new StudentService();
+    assertEquals(false, studentService.validInput("abc", "12", "abs scs", "12", "A C D"));
+  }
+
+  @Test
+  void checkIfRollNoAlreadyExistWhenAddingStudent() {
+    StudentService studentService = new StudentService();
+    studentService.processInput("abc", "12", "x y z", "12", "A B C D");
+    studentService.processInput("def", "12", "m n o", "22", "C D E F");
+    assertEquals(1, studentService.studentList.size());
```

Comment:
shouldn't the size value be 2?

Reply:
Since both has same rollNo only first entry will be stored second will be discarded.

Reply:
ohhk correct